### PR TITLE
Metadata192 lower

### PR DESCRIPTION
**Pull request by brantpeery**
Does a deep compare of metadata to determine if the metadata was saved by 19.3 on the first call. If the compare fails, it will try to save the metadata using the 19.2 method. 
I left in the requestSystemVersion() method for later use.

### CODE REVIEW

File: pkg/venafi/tpp/connector.go [hunk 1]
```diff
@@ -258,6 +258,120 @@ func wrapAltNames(req *certificate.Request) (items []sanItem) {
...
+		return nil, fmt.Errorf("%s", err)
+	}
+	return response.Data, nil
+}
+
+//RequestSystemVersion returns the TPP system version of the connector context
+func (c *Connector) requestSystemVersion() (ver string, err error) {
+	statusCode, status, body, err := c.request("GET", urlResourceSystemStatusVersion, "")
+	if err != nil {
+		return "", fmt.Errorf("%s", err)
```
Comment (mr-tron) [inline pkg/venafi/tpp/connector.go]: just return err

File: pkg/venafi/tpp/connector.go [hunk 2]
```diff
@@ -258,6 +258,120 @@ func wrapAltNames(req *certificate.Request) (items []sanItem) {
...
+	}
+	//Put in hint for authentication scope 'configuration'
+	switch statusCode {
+	case 200:
+		break
```
Comment (mr-tron) [inline pkg/venafi/tpp/connector.go]: unnecessary break

File: pkg/venafi/tpp/connector.go [hunk 3]
```diff
@@ -258,6 +258,120 @@ func wrapAltNames(req *certificate.Request) (items []sanItem) {
...
+	case 401:
+		return "", fmt.Errorf("http status code '%s' was returned by the server. Hint: OAuth scope 'configuration' is required when using custom fields", status)
+	default:
+		return "", fmt.Errorf("Unexpected http status code while fetching TPP version. %s", status)
+	}
+	type responseType struct{ Version string }
+	response := responseType{}
```
Comment (mr-tron) [inline pkg/venafi/tpp/connector.go]: you can use anonymous type 
```
var response struct{
     Version string
}
```

File: pkg/venafi/tpp/connector.go [hunk 4]
```diff
@@ -258,6 +258,120 @@ func wrapAltNames(req *certificate.Request) (items []sanItem) {
...
+	err = json.Unmarshal(body, &response)
+	if err != nil {
+		return "", fmt.Errorf("%s", err)
+	}
+	return response.Version, nil
+}
+
+//SetCertificateMetadata submits the metadata to TPP for storage returning the lock status of the metadata stored
+func (c *Connector) setCertificateMetadata(metadataRequest metadataSetRequest) (locked bool, err error) {
+	if len(metadataRequest.DN) < 1 {
```
Comment (mr-tron) [inline pkg/venafi/tpp/connector.go]: == 0

File: pkg/venafi/tpp/connector.go [hunk 5]
```diff
@@ -258,6 +258,120 @@ func wrapAltNames(req *certificate.Request) (items []sanItem) {
...
+		return false, fmt.Errorf("DN must be provided to setCertificateMetaData")
+	}
+	if len(metadataRequest.GuidData) < 1 && metadataRequest.KeepExisting {
```
Comment (mr-tron) [inline pkg/venafi/tpp/connector.go]: == 0

File: pkg/venafi/tpp/connector.go [hunk 6]
```diff
@@ -258,6 +258,120 @@ func wrapAltNames(req *certificate.Request) (items []sanItem) {
...
+		return false, nil
+	} //Not an error, but there is nothing to do
+
+	statusCode, status, body, err := c.request("POST", urlResourceMetadataSet, metadataRequest)
+	if err != nil {
+		return false, fmt.Errorf("%s", err)
```
Comment (mr-tron) [inline pkg/venafi/tpp/connector.go]: just `return false, err`

File: pkg/venafi/tpp/connector.go [hunk 7]
```diff
@@ -395,6 +509,58 @@ func (c *Connector) RequestCertificate(req *certificate.Request) (requestID stri
...
+		//the 19.2 WebSDK calls
+		metadataItems, err := c.requestMetadataItems(requestID)
+		if err != nil {
+			return "", fmt.Errorf("%s", err)
+		}
+
+		//Deep compare the request metadata to the fetched metadata
+		var allItemsFound = true
+	OUTER:
+		for _, cf := range tppCertificateRequest.CustomFields {
```
Comment (mr-tron) [inline pkg/venafi/tpp/connector.go]: Please rewrite this code. Remove labeled cycles. 
4 nested cycles can replaced with search in prepared map,  helper functions.

File: pkg/venafi/tpp/tpp.go [hunk 8]
```diff
@@ -209,6 +209,86 @@ type policyRequest struct {
...
+	RenderReadOnly    bool     `json:",omitempty"`
+	Type              int      `json:",omitempty"`
+}
+type metadataKeyValueSet struct {
+	Key   metadataItem `json:",omitempty"`
+	Value []string     `json:",omitempty"`
+}
+type metadataResultCode int
+
+var metadataResultCodesMap = map[string]metadataResultCode{
```
Comment (mr-tron) [inline pkg/venafi/tpp/tpp.go]: unused variable

File: pkg/venafi/tpp/tpp.go [hunk 9]
```diff
@@ -209,6 +209,86 @@ type policyRequest struct {
...
+	RegularExpression string   `json:",omitempty"`
+	RenderHidden      bool     `json:",omitempty"`
+	RenderReadOnly    bool     `json:",omitempty"`
+	Type              int      `json:",omitempty"`
+}
+type metadataKeyValueSet struct {
+	Key   metadataItem `json:",omitempty"`
+	Value []string     `json:",omitempty"`
+}
+type metadataResultCode int
```
Comment (mr-tron) [inline pkg/venafi/tpp/tpp.go]: all fields with type metadataResultCode is unused

**Comment by munnerz** [comment]
cc

File: pkg/venafi/tpp/connector.go (hunk 7 above)
Comment (brantpeery) [reply pkg/venafi/tpp/connector.go]: I am not the best a Go maps. Can you help me understand how that is cleaner and faster?

**Comment by mr-tron** [comment]
I fixed all problems that I found in https://github.com/Venafi/vcert/tree/pull/87
You can merge it to your pull request or I can close your request and make new.
All automatic tests passed for my branch `pull/87`

**Comment by mr-tron** [comment]
https://github.com/Venafi/vcert/pull/94

**Pull request closed without merging**